pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,210 @@
+	velerov1client "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned/typed/velero/v1"
+	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/persistence"
+	"github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt"
+)
+
+type backupStorageLocationController struct {
+	*genericController
+
+	namespace                    string
+	defaultBackupLocation        string
+	defaultStoreValidationPeriod time.Duration
+	backupLocationClient         velerov1client.BackupStorageLocationsGetter
+	backupStorageLocationLister  velerov1listers.BackupStorageLocationLister
+	newPluginManager             func(logrus.FieldLogger) clientmgmt.Manager
+	newBackupStore               func(*velerov1api.BackupStorageLocation, persistence.ObjectStoreGetter, logrus.FieldLogger) (persistence.BackupStore, error)
+}
+
+func NewBackupStorageLocationController(
+	namespace string,
+	defaultBackupLocation string,
+	defaultStoreValidationPeriod time.Duration,
+	backupLocationClient velerov1client.BackupStorageLocationsGetter,
+	backupStorageLocationLister velerov1listers.BackupStorageLocationLister,
+	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
+	logger logrus.FieldLogger,
+) Interface {
+	if defaultStoreValidationPeriod <= 0 {
+		defaultStoreValidationPeriod = time.Minute
+	}
+
+	logger.Infof("Backup location store validation period is %v", defaultStoreValidationPeriod)
+
+	c := backupStorageLocationController{
+		genericController:            newGenericController("backup-storage-location", logger),
+		namespace:                    namespace,
+		defaultBackupLocation:        defaultBackupLocation,
+		defaultStoreValidationPeriod: defaultStoreValidationPeriod,
+		backupLocationClient:         backupLocationClient,
+		backupStorageLocationLister:  backupStorageLocationLister,
+
+		// use variables to refer to these functions so they can be
+		// replaced with fakes for testing.
+		newPluginManager: newPluginManager,
+		newBackupStore:   persistence.NewObjectBackupStore,
+	}
+
+	c.resyncFunc = c.run
+	c.resyncPeriod = 30 * time.Second
+
+	return c
+}
+
+func (c *backupStorageLocationController) run() {
+	c.logger.Info("Checking for existing locations ready to be validated; there needs to be at least 1 location that is available")
+
+	locations, err := c.backupStorageLocationLister.BackupStorageLocations(c.namespace).List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(err).Error("Error listing locations, at least one available location is required")
+		return
+	}
+
+	if len(locations) == 0 {
+		c.logger.Error("No locations found, at least one available location is required")
+		return
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+
+	var unavailable []string
+	log := c.logger
+	for _, location := range locations {
+		locationName := location.Name
+		log = c.logger.WithField("backupLocation", locationName)
+
```

Comment:
Is `location.Spec.StoreValidationPeriod == nil` also a condition where we should skip validation?

Reply:
No. Only a value of zero is meant to skip validation. Any other value will set the frequency to that value (obvs), and nil means they didn't set it, which then the frequency will be set to the server value, either one set by the user, or the default. Pls be sure to see my comment up above.